tests/15_Versioning/VersionManagerTest.php

```diff
@@ -275,6 +282,113 @@ public function testRestorePendingChanges()
+        $this->vm->checkout($nodePath);
+
+        $node = $this->session->getNode($nodePath);
+        $node->setProperty('bar', 'value');
+
+        $this->session->save();
+
+        $this->assertTrue($node->hasProperty('bar'));
+
+        $this->vm->restore(true, $version);
+        $node = $this->session->getNode($nodePath);
+
+        $this->assertFalse($node->hasProperty('bar'));
+    }
+
+    public function testRestoreIsCheckedIn()
+    {
+        $nodePath = '/tests_version_base/versioned';
+        $version = $this->vm->checkpoint($nodePath);
+
+        $this->vm->checkout($nodePath);
+        $this->assertTrue($this->vm->isCheckedOut($nodePath));
+
+        $this->vm->restore(true, $version);
+        $this->assertFalse($this->vm->isCheckedOut($nodePath));
+    }
+
+    public function testRestoreBaseProperties()
+    {
+        // TODO also check for primary node type once it can be changed
+
+        $nodePath = '/tests_version_base/versioned';
+        $version = $this->vm->checkpoint($nodePath);
+        $this->vm->checkout($nodePath);
+
+        $node = $this->session->getNode($nodePath);
+        $node->addMixin('mix:created');
+
+        $this->session->save();
+
+        $node = $this->session->getNode($nodePath);
+        $this->assertContains('mix:created', $node->getPropertyValue('jcr:mixinTypes'));
+
+        $this->vm->restore(true, $version);
+
+        $node = $this->session->getNode($nodePath);
+        $this->assertEquals(array('mix:versionable'), $node->getPropertyValue('jcr:mixinTypes'));
+    }
+
+    public function testRestoreWithChildren()
+    {
+        $nodePath = '/tests_version_base/versioned';
+        $this->vm->checkout($nodePath);
+        $node = $this->session->getNode($nodePath);
+        $childNode1 = $node->addNode('childNode1');
+        $childNode1->setProperty('foo', 'child1');
+        $childNode2 = $node->addNode('childNode2');
+        $childNode2->setProperty('foo', 'child2');
+
+        $this->session->save();
+        $version = $this->vm->checkin($nodePath);
+
+        $this->assertCount(2, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode1')->getPropertyValue('foo'));
+        $this->assertEquals('child2', $node->getNode('childNode2')->getPropertyValue('foo'));
+
+        $this->vm->checkout($nodePath);
+
+        $childNode1->remove();
+        $childNode2->setProperty('foo', 'child1');
+
+        $this->session->save();
+
+        $this->assertCount(1, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode2')->getPropertyValue('foo'));
+
+        $this->vm->restore(false, $version);
+        $node = $this->session->getNode($nodePath);
+
+        $this->assertCount(2, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode1')->getPropertyValue('foo'));
+        $this->assertEquals('child2', $node->getNode('childNode2')->getPropertyValue('foo'));
```

Comment:
This is interesting, it looks to me like this is an error in the jackalope jackrabbit transport layer, since jackrabbit itself (I have tested it via the cli in jackrabbit-standalone.jar) behaves as expected, as you can see in the following script:

```
[/] > addnode test

elapsed time: 25 ms.

[/*] > addmix

addmix     addmixin
[/*] > addmixin mix:versionable test

elapsed time: 2 ms.

[/*] > cd test

elapsed time: 0 ms.

[/test*] > addnode childnode1

elapsed time: 1 ms.

[/test*] > cd childnode1

elapsed time: 0 ms.

[/test/childnode1*] > setprop

setprop               setproperty           setpropertyfromfile   setpropfromfile
[/test/childnode1*] > setprop foo child1

elapsed time: 2 ms.

[/test/childnode1*] > cd ..

elapsed time: 0 ms.

[/test*] > addnode childnode2

elapsed time: 1 ms.

[/test*] > cd childnode

childnode1   childnode2
[/test*] > cd childnode2

elapsed time: 0 ms.

[/test/childnode2*] > setprop

setprop               setproperty           setpropertyfromfile   setpropfromfile
[/test/childnode2*] > setprop foo child2

elapsed time: 0 ms.

[/test/childnode2*] > cd ..

elapsed time: 0 ms.

[/test*] > save

elapsed time: 74 ms.

[/test] > checkin

elapsed time: 118 ms.

[/test] > cd childnode

childnode1   childnode2
[/test] > cd childnode1

elapsed time: 0 ms.

[/test/childnode1] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
jcr:primaryType                false     Name      15       nt:unstructured
foo                            false     String    6        child1


total
2 nodes

elapsed time: 3 ms.

[/test/childnode1] > cd ../childnode2

elapsed time: 0 ms.

[/test/childnode2] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
foo                            false     String    6        child2
jcr:primaryType                false     Name      15       nt:unstructured


total
2 nodes

elapsed time: 1 ms.

[/test/childnode2] > cd ..

elapsed time: 0 ms.

[/test] > checkout

elapsed time: 19 ms.

[/test] > rm childnode

childnode1   childnode2
[/test] > rm childnode1

elapsed time: 2 ms.

[/test*] > cd childnode2

elapsed time: 0 ms.

[/test/childnode2*] > setprop

setprop               setproperty           setpropertyfromfile   setpropfromfile
[/test/childnode2*] > setprop foo child1

elapsed time: 1 ms.

[/test/childnode2*] > save

elapsed time: 18 ms.

[/test/childnode2] > cd ..

elapsed time: 0 ms.

[/test] > ls
name                           type            node      new       modified
------------------------------ --------------- --------- --------- ---------
childnode2                     nt:unstructured true      false     false
jcr:isCheckedOut               Boolean         false     false     false
jcr:versionHistory             Reference       false     false     false
jcr:mixinTypes                 Name            false     false     false
jcr:predecessors               Reference       false     false     false
jcr:baseVersion                Reference       false     false     false
jcr:uuid                       String          false     false     false
jcr:primaryType                Name            false     false     false

total

elapsed time: 5 ms.

[/test] > cd childnode2

elapsed time: 1 ms.

[/test/childnode2] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
foo                            false     String    6        child1
jcr:primaryType                false     Name      15       nt:unstructured


total
2 nodes

elapsed time: 1 ms.

[/test/childnode2] > cd ..

elapsed time: 0 ms.

[/test] > lsversions
version              labels
-------------------- --------------------------------------------------
jcr:rootVersion
1.0

elapsed time: 32 ms.

[/test] > restore 1.0

elapsed time: 32 ms.

[/test] > ls
name                           type            node      new       modified
------------------------------ --------------- --------- --------- ---------
childnode1                     nt:unstructured true      false     false
childnode2                     nt:unstructured true      false     false
jcr:isCheckedOut               Boolean         false     false     false
jcr:versionHistory             Reference       false     false     false
jcr:mixinTypes                 Name            false     false     false
jcr:predecessors               Reference       false     false     false
jcr:baseVersion                Reference       false     false     false
jcr:uuid                       String          false     false     false
jcr:primaryType                Name            false     false     false

total

elapsed time: 13 ms.

[/test] > cd childnode

childnode1   childnode2
[/test] > cd childnode1

elapsed time: 0 ms.

[/test/childnode1] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
jcr:primaryType                false     Name      15       nt:unstructured
foo                            false     String    6        child1
jcr:mixinTypes                 true[0]   Name      0


total
3 nodes

elapsed time: 1 ms.

[/test/childnode1] > cd ../childnode2

elapsed time: 1 ms.

[/test/childnode2] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
jcr:mixinTypes                 true[0]   Name      0
foo                            false     String    6        child2
jcr:primaryType                false     Name      15       nt:unstructured


total
3 nodes

elapsed time: 2 ms.
```